coderd/notifications/metrics.go

```diff
@@ -0,0 +1,71 @@
+package notifications
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+type Metrics struct {
+	DispatchedCount  *prometheus.CounterVec
+	TempFailureCount *prometheus.CounterVec
+	PermFailureCount *prometheus.CounterVec
+	RetryCount       *prometheus.CounterVec
+
+	QueuedSeconds *prometheus.HistogramVec
+
+	DispatcherSendSeconds *prometheus.HistogramVec
+
+	PendingUpdates prometheus.Gauge
+}
+
+const (
+	ns        = "coderd"
+	subsystem = "notifications"
+
+	LabelMethod     = "method"
+	LabelTemplateID = "template_id"
+)
+
+func NewMetrics(reg prometheus.Registerer) *Metrics {
+	return &Metrics{
+		DispatchedCount: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Name: "dispatched_count", Namespace: ns, Subsystem: subsystem,
+			Help: "The count of notifications successfully dispatched.",
+		}, []string{LabelMethod, LabelTemplateID}),
+		TempFailureCount: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Name: "temporary_failures_count", Namespace: ns, Subsystem: subsystem,
+			Help: "The count of notifications which failed but have retry attempts remaining.",
+		}, []string{LabelMethod, LabelTemplateID}),
+		PermFailureCount: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Name: "permanent_failures_count", Namespace: ns, Subsystem: subsystem,
+			Help: "The count of notifications which failed and have exceeded their retry attempts.",
+		}, []string{LabelMethod, LabelTemplateID}),
+		RetryCount: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Name: "retry_count", Namespace: ns, Subsystem: subsystem,
+			Help: "The count of notification dispatch retry attempts.",
+		}, []string{LabelMethod, LabelTemplateID}),
+
+		// Aggregating on LabelTemplateID as well would cause a cardinality explosion.
+		QueuedSeconds: promauto.With(reg).NewHistogramVec(prometheus.HistogramOpts{
+			Name: "queued_seconds", Namespace: ns, Subsystem: subsystem,
+			Buckets: []float64{1, 5, 15, 30, 60, 120, 300, 600, 3600, 86400},
```

Comment:
We say this should be less than the fetch interval, which is 15s --- so I think there should be more buckets <15s.  In fact, it should be more like 15/(n+1) where n is the number of Coderd replicas, assuming that notifications usually get picked up by the next fetch.  Seeing the average move up toward 15s is way more important than the distinction between 10minutes, 1 hour and 1 day, all of which indicate that notifications are just blocked entirely.